### ISSUE TITLE
List component does not update when filter property is updated

### ISSUE THREAD

**Issue by kara-todd** [marmelab/react-admin#2923]
**What you were expecting:**
When dynamically updating the List `filter` property using a redux action the list data should immediately be updated.

**What happened instead:**
The list data is delayed in updating creating a confusing UX.

**Steps to reproduce:**
* [CodeSandbox](https://codesandbox.io/s/6x9ljr1q6r) and [demo](https://6x9ljr1q6r.codesandbox.io/#/users)
1. Sort by ID to generate query string params
2. Select "Group A" from the dropdown (no change to list)
3. Select "Group B" from the dropdown  (list is now filtered by "Group A")

If query string params are not present an error is thrown to the browser  console `Warning: Missing translation for key: "Cannot read property 'toLowerCase' of null"`.

**Other information:**
I think this is related to the `shouldUpdateComponent` function in the [ListController](https://github.com/marmelab/react-admin/blob/master/packages/ra-core/src/controller/ListController.tsx#L220) it seems like perhaps just adding the filter property to this list could fix the issue. However, I wasn't sure if it was intentionally excluded for performance reasons...

I think [this stack overflow issue](https://stackoverflow.com/questions/51378884/how-to-use-a-custom-reducers-state-as-a-permanent-filter-in-a-list) has the same root cause.

**Environment**

I have replicated this with two environments:

* React-admin version: 2.4.0, React version: 16.5.2
* React-admin version: 2.7.2, React version: 16.8.3

**Comment by kara-todd** [comment]
Our reason for wanting this is that we are using the `filter` parameter as a sort of "system imposed" filter that the user cannot directly change. Regular list filter parameters are still exposed to users and can be modified by them.

**Comment by djhi** [comment]
Thanks for the detailed report. This is indeed the same issue as in the StackOverflow question. Would you mind openning a PR?

**Comment by kara-todd** [comment]
Hi @djhi I'd love to... unfortunately, I'm having issues getting the test suite to pass even after I cloned down the repo directly. :(

I took the following steps:
1. $ `git clone https://github.com/marmelab/react-admin .`
2. $ `make install` (or yarn install)
3. $ `make test-unit`

All of my `packages/ra-ui-materialui/*` tests fail with `Cannot find module 'ra-core'`

Did I miss a step here?

**Comment by djhi** [comment]
Ah you need to build `ra-core` once in order to unit test the other packages. Try running `make build-ra-core`

**Comment by kara-todd** [comment]
@djhi I ran `make build` first and all my unit tests passed. 👍

I've opened a PR. Some of my end-to-end tests are timing out, but they shouldn't be impacted by the changes.

**Comment by fzaninotto** [comment]
for "system imposed" filter, use the `permanentFilter` prop

**Comment by kara-todd** [comment]
@fzaninotto I created a [codesandbox fork](https://codesandbox.io/s/p2mj72wo80) using the `permanentFilter` prop but it also doesn't seem to be working as I would expect. Did I miss a step?

**Comment by fzaninotto** [comment]
sorry, it's called `filter`. https://marmelab.com/react-admin/List.html#permanent-filter

**Comment by Gonusi** [comment]
Using React Admin v. 2.92, I tried dynamically changing the ```filter``` prop on a List component, however ran into a few issues.

1. When dynamically changing ```filter``` prop at runtime, sort order and pagination become ```null```. I presume this has to do with them not being in the URL. I can temporarily fix this by setting the default ```sort``` and ```perPage``` props on the List component.
2. The first time filter prop is set, it does not reacth data provider. Then when you turn them off, data provider receives correct filter from previous step, even it "undefined" passed as filter prop. So the sequence is like this:
    1. <List filter={id:'1'} /> -> Not filtered
    2. <List filter={undefined} -> Data provider successfully has filters props
    2. <List filter={id: '1} /> -> Not filtered

The behaviour can be seen using this wrapper of List component:

```
const BoList = ({ userIdFilterValue, ...props }) => {
	const [isEnabled, setIsEnabled] = useState(false);
	return (
		<>
			<div onClick={setIsEnabled(!isEnabled)}>Toggle Filter</div>
			<List
				{...props}
				filter={isEnabled ? { userId: '8f50eacf-758d-4848-b9f0-c238b78b10d6' } : undefined}
			/>
		</>
	);
};
```

**Comment by rinde** [comment]
Any updates on this issue?

I'm experiencing the same behavior as @Gonusi: the filter always lags one step behind. I am integrating with a calendar component and I want the filter to adjust to the period that is visible in the calendar. What I'm observing though is that the data that is loaded is the data of the previous period.

**Comment by nicgirault** [comment]
Same issue here trying to filter data according to the selected locale. Any plan to fix the issue?

**Comment by kara-todd** [comment]
Since there has been some recent comments here. I opened a PR a while back (#2926) but it's since fallen behind.

**Comment by djhi** [comment]
Sorry about this. It has been fixed by #3308 and I forgot to keep you updated.
